File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/IntegrationTools.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Net.Http
open System.Text
open System.Text.Json
open System.Threading.Tasks

module IntegrationTools =

    let private httpClient = new HttpClient()

    let sendWebhookAsync (argsJson: string) : Task<string> =
        async {
            try
                let doc = JsonDocument.Parse argsJson
                use _ = doc
                let root = doc.RootElement
                match root.TryGetProperty("url") with
                | true, urlProp ->
                    let url = urlProp.GetString()
                    let methodStr = if root.TryGetProperty("method") <> (false, Unchecked.defaultof<_>) then root.GetProperty("method").GetString().ToUpper() else "POST"
                    let payload = if root.TryGetProperty("payload") <> (false, Unchecked.defaultof<_>) then root.GetProperty("payload").ToString() else "{}"

                    use request = new HttpRequestMessage(HttpMethod(methodStr), url)
                    request.Content <- new StringContent(payload, Encoding.UTF8, "application/json")

                    if root.TryGetProperty("headers") <> (false, Unchecked.defaultof<_>) && root.GetProperty("headers").ValueKind = JsonValueKind.Object then
                        for prop in root.GetProperty("headers").EnumerateObject() do
                            request.Headers.TryAddWithoutValidation(prop.Name, prop.Value.ToString()) |> ignore

                    let! response = httpClient.SendAsync(request) |> Async.AwaitTask
                    let! body = response.Content.ReadAsStringAsync() |> Async.AwaitTask
                    let truncated = ToolSecurity.truncateOutput 5000 body
                    return sprintf "Webhook HTTP %d (%s):\n%s" (int response.StatusCode) response.ReasonPhrase truncated
                | false, _ -> return "Error: Missing 'url' argument."
            with ex -> return sprintf "Error sending webhook: %s" ex.Message
        } |> Async.StartAsTask

File: implementations/csharp/src/Skight.AgentPlatform/IntegrationTools.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace Skight.AgentPlatform
{
    public static class IntegrationTools
    {
        private static readonly HttpClient HttpClient = new HttpClient();

        public static async Task<string> SendWebhookAsync(string argsJson)
        {
            try
            {
                using var doc = JsonDocument.Parse(argsJson);
                var root = doc.RootElement;
                if (!root.TryGetProperty("url", out var urlProp))
                {
                    return "Error: Missing 'url' argument.";
                }

                var url = urlProp.GetString()!;
                var method = root.TryGetProperty("method", out var mProp) ? mProp.GetString()!.ToUpper() : "POST";
                var payload = root.TryGetProperty("payload", out var pProp) ? pProp.ToString() : "{}";

                using var request = new HttpRequestMessage(new HttpMethod(method), url);
                request.Content = new StringContent(payload, Encoding.UTF8, "application/json");

                if (root.TryGetProperty("headers", out var headersProp) && headersProp.ValueKind == JsonValueKind.Object)
                {
                    foreach (var header in headersProp.EnumerateObject())
                    {
                        request.Headers.TryAddWithoutValidation(header.Name, header.Value.ToString());
                    }
                }

                var response = await HttpClient.SendAsync(request);
                var body = await response.Content.ReadAsStringAsync();
                var truncatedBody = ToolSecurity.TruncateOutput(5000, body);

                return $"Webhook HTTP {(int)response.StatusCode} ({response.ReasonPhrase}):\n{truncatedBody}";
            }
            catch (Exception ex)
            {
                return $"Error sending webhook: {ex.Message}";
            }
        }
    }
}
